FATE Threads Changes

kevinrr888 opened the pull request:
This PR:
- Changes FATE from having one pool of workers per FATE type (META and USER) to having a set of pools per FATE type which each work on their own subset of FATE operations:
	- This is accomplished through a new class `FateExecutor`. Each FATE has a set of `FateExecutor`s. Each `FateExecutor` is assigned to one set of FATE operations, has a `WorkFinder` thread, and a pool of `TransactionRunner` threads. The `WorkFinder` finds transactions this `FateExecutor` can work on, and assigns them to the `TransactionRunner`s which work on them.
	- `Fate` class now only handles the `DeadReservationCleaner` thread and the `FatePoolsWatcher` thread (previously `FatePoolWatcher`). The `DeadReservationCleaner` is unchanged. `FatePoolsWatcher` now oversees all the `FateExecutor`s/pools. The thread periodically checks for changes to the config, resizing any pools as necessary, safely stopping (if a thread is working on a transaction, it is not terminated until it finishes that work) any `FateExecutor`s that are invalidated by config changes (i.e., its assigned set of FATE operations is no longer present in the config), and starting any new/replacement `FateExecutor`s needed. The `FatePoolsWatcher` also keeps track of the number of idle threads to warn the user if a pool is experiencing a high load and should have its pool size increased (this functionality is unchanged from before, only difference is now a slightly different message is logged when warning the user).
	- The set of pools is determined by the new `MANAGER_USER_FATE_CONFIG`/`MANAGER_META_FATE_CONFIG` properties (replaces `MANAGER_FATE_THREADPOOL_SIZE`).
	- Another difference from this change is now USER and META FATEs can be configured differently whereas previously they shared the same property (`MANAGER_FATE_THREADPOOL_SIZE`)
	- These properties are json where the keys are a set of FATE operations and values are a pool size.
	- The properties are ensured to be valid json, pool size > 0, contain all FATE operations, contain no invalid FATE operations, and have no duplicate FATE operations
	- The FATE operations can be split across any number of pools of any size.
- Tests which used the `MANAGER_FATE_THREADPOOL_SIZE` have now been changed to `MANAGER_USER_FATE_CONFIG`/`MANAGER_META_FATE_CONFIG` properties. To keep testing functionality the same, the properties set are json of 1 key/value where the key is all FATE ops and the value is the value originally used in the test.
- `FatePoolResizeIT`/`MetaFatePoolResizeIT`/`UserFatePoolResizeIT` have been renamed to `FatePoolsWatcherIT`/`MetaFatePoolsWatcherIT`/`UserFatePoolsWatcherIT` since they now test other functionality of the `FatePoolsWatcher` not just pool resizing. The tests have also been completely changed since resizing has completely changed.

I also ensured all FATE tests still pass and sunny day still passes. I have not yet run the full set of ITs for this.

closes #5130

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/FateExecutor.java:
```diff
@@ -0,0 +1,477 @@
+  private final ConcurrentLinkedQueue<Integer> idleCountHistory = new ConcurrentLinkedQueue<>();
+
+  public FateExecutor(Fate<T> fate, T environment, Set<Fate.FateOperation> fateOps, int poolSize) {
+    final String operatesOn = fate.getStore().type().name().toLowerCase() + "."
+        + fateOps.stream().map(fo -> fo.name().toLowerCase()).collect(Collectors.joining("."));
+    final String transactionRunnerPoolName =
+        ThreadPoolNames.MANAGER_FATE_POOL_PREFIX.poolName + operatesOn;
+    final String workFinderThreadName = "fate.work.finder." + operatesOn;
```
Example pool name: `accumulo.pool.manager.fate.meta.table_split.system_split.table_delete_range.table_merge`
This is long, but descriptive

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/user/UserFateStore.java:
```diff
@@ -256,21 +256,24 @@ protected Stream<FateIdStatus> getTransactions(EnumSet<TStatus> statuses) {
         SortedMap<Key,Value> rowMap;
         TStatus status = TStatus.UNKNOWN;
         FateReservation reservation = null;
+        Fate.FateOperation fateOp = null;
```
My impl includes rename of `txName` to `fateOp` in the places where I write new code. Ignore the discrepancy in naming: these are the same and will be consolidated in a later PR via https://github.com/apache/accumulo/issues/5230.

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/conf/Property.java:
```diff
@@ -429,15 +429,34 @@ public enum Property {
-      "The number of threads used to run fault-tolerant executions (FATE)."
-          + " These are primarily table operations like merge.",
-      "1.4.3"),
+  MANAGER_USER_FATE_CONFIG("manager.user.fate.config", "{"
+      + "\"TABLE_CREATE,TABLE_DELETE,TABLE_RENAME,TABLE_ONLINE,TABLE_OFFLINE,NAMESPACE_CREATE,NAMESPACE_DELETE,NAMESPACE_RENAME,TABLE_TABLET_AVAILABILITY,SHUTDOWN_TSERVER\": 1,"
+      + "\"TABLE_BULK_IMPORT2\": 2,"
+      + "\"TABLE_COMPACT,TABLE_CANCEL_COMPACT,COMMIT_COMPACTION\": 4,"
+      + "\"TABLE_MERGE,TABLE_DELETE_RANGE,TABLE_SPLIT,SYSTEM_SPLIT,TABLE_CLONE,TABLE_IMPORT,TABLE_EXPORT\": 2"
+      + "}", PropertyType.USER_FATE_CONFIG,
+      "The number of threads used to run user-initiated fault-tolerant "
+          + "executions (FATE). These are primarily table operations like merge. Each key/value "
+          + "of the provided JSON corresponds to one thread pool. Each key is a list of one or "
+          + "more FATE operations and each value is the number of threads that will be assigned "
+          + "to the pool.",
+      "4.0.0"),
+  MANAGER_META_FATE_CONFIG("manager.meta.fate.config",
+      "{\"TABLE_COMPACT,TABLE_CANCEL_COMPACT,COMMIT_COMPACTION\": 4,"
+          + "\"TABLE_MERGE,TABLE_DELETE_RANGE,TABLE_SPLIT,SYSTEM_SPLIT\": 2}",
+      PropertyType.META_FATE_CONFIG,
+      "The number of threads used to run system-initiated fault-tolerant "
+          + "executions (FATE). These are primarily table operations like merge. Each key/value "
+          + "of the provided JSON corresponds to one thread pool. Each key is a list of one or "
+          + "more FATE operations and each value is the number of threads that will be assigned "
+          + "to the pool.",
+      "4.0.0"),
```
Several comments here:
- This is my best attempt as to what fate ops are valid META fate ops. I may have excluded some that should have been included or vice-versa.
- What is the proper way to deprecate the old MANAGER_FATE_THREADPOOL_SIZE property
- These default pool sizes and default pools probably aren't optimal. Input into how the FATE ops should be split and how many threads for each pool would be appreciated. This was just my best attempt

kevinrr888 replied:
I marked as deprecated in 4.0 and log a warning if the property is set mentioning the replacement props.

kevinrr888 commented on test/src/main/java/org/apache/accumulo/test/fate/MultipleStoresIT.java:
```diff
@@ -249,30 +249,32 @@ private void testMultipleFateInstances(TestStoreFactory<SleepingTestEnv> testSto
-      } else {
-        fateId = fate2.startTransaction();
-        fate2.seedTransaction(TEST_FATE_OP, fateId, new SleepingTestRepo(), true, "test");
+    try {
```
I made shutdown() in a finally block for the tests in this class to avoid failures in one test affecting the others. Should have already existed in the test.

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/Fate.java:
```diff
@@ -144,260 +140,175 @@ public TFateOperation toThrift() {
+          final var poolName = fateExecutor.getPoolName();
+          final var runningTxRunners = fateExecutor.getRunningTxRunners();
+          final int configured = poolConfigs.get(fateExecutor.getFateOps());
+          ThreadPools.resizePool(pool, () -> configured, poolName);
```
I have not yet explored changing this into a cached thread pool (discussed in https://github.com/apache/accumulo/pull/5263#discussion_r1917457670 thread). The changes were already pretty large and didn't want to introduce more. Writing this as a reminder to do as a potential follow on.

dlmarion reviewed (commented):
Looks good.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/Fate.java:
```diff
@@ -413,111 +324,81 @@ public void run() {
+  protected Map<Set<FateOperation>,Integer> getPoolConfigurations(AccumuloConfiguration conf) {
+    Map<Set<FateOperation>,Integer> poolConfigs = new HashMap<>();
+    final var json = JsonParser.parseString(conf.get(getFateConfigProp())).getAsJsonObject();
+
```
The code in PropertyType validates that all FateOperatiopns are present in the config that there is no duplication.  Not sure where this should go, but in the case where those conditions are not met will it get to this code?

Also it would be nice to log somewhere what the problem, if that is not being done.  For example log that  COMMIT_COMPACTION was not seen in the config, or log that SYSTEM_SPLIT was seen twice in the config.  Not sure where that should be done, was thinking that maybe it could be done here.

kevinrr888 replied:
If PropertyType acts how I expect, it should not reach this code. My interpretation was that PropertyType is how the property is tested for validity before ever being set ensuring an invalid prop can never be set. If this is not how property type works, it may reach this code. (I'm not sure what other purpose PropertyType would serve if not this, though).

Added logs to PropertyType to explain the reason for failure-to-set.

keith-turner reviewed (commented):
Only partially looked at this, will look more later.

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/conf/PropertyType.java:
```diff
@@ -393,4 +414,86 @@ public static IntStream parse(String portRange) {
+    private static final Logger log = LoggerFactory.getLogger(FateThreadPoolSize.class);
+
+    @Override
+    public boolean test(String s) {
+      log.warn(
+          "The manager fate thread pool size property is no longer used. See the {} and {} for "
+              + "the replacements to this property.",
+          ValidUserFateConfig.NAME, ValidMetaFateConfig.NAME);
+      return true;
```
Would be nice to print the actual property but I don't think it's possible since Property uses PropertyType to construct each Property. Accessing Property here may introduce circular dependencies. This is safe and probably good enough.

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/zookeeper/MetaFateStore.java:
```diff
@@ -584,6 +584,16 @@ public TStatus getStatus() {
+              return Optional.empty();
+            } else {
+              return Optional.of((Fate.FateOperation) fateOp);
+            }
```
Ah yeah that works. I thought casting null results in NPE, but it doesn't.

keith-turner commented on test/src/main/java/org/apache/accumulo/test/fate/FatePoolsWatcherIT.java:
```diff
@@ -0,0 +1,477 @@
+/**
+ * Tests the functionality of the FATE pools watcher task
+ */
+public abstract class FatePoolsWatcherIT extends SharedMiniClusterBase
```
These are really nice test.  

Not sure if its covered, another test that may be useful is one that ensures things are being partitioned correctly among the pools.  

The following is a rough stab at doing this.

 1. Create FateExectuor_1 w/ 2 threads
 2. Create FateExectuor_2 w/ 3 threads
 3. Create FateExectuor_3 w/ 4 threads
 4. Create FateExectuor_4 w/ 5 threads
 5. Seed 5 things to run on FateExectuor_1
 6. Seed 6 things to run on FateExectuor_2
 7. Seed 1 thing to run on FateExectuor_3
 8. Seed 4 things to run on FateExectuor_4
 
Should eventually see 2 running on FateExectuor_1, 3 running on FateExectuor_2, 1 running on FateExectuor_3, and 4 running on FateExectuor_4.  So for each executor should see the `min(configuredThreads, seededTask)` actually running.  Giving some executors more work than they have threads and some less at the same time. Hoping this would pick up on some cases where an executor snags some work its not supposed to.  Tried to pick number that made the inputs to min() unique and the result of min() unique.

The existing test seem to create the exact number of task as threads in most cases. When the existing test create more task that threads seems like its only for one executor.  Was attempting to have multiple executors w/ different threads and different numbers of task that differ from the number of threads in for the target executor.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/conf/Property.java:
```diff
@@ -429,15 +429,41 @@ public enum Property {
+  MANAGER_FATE_USER_CONFIG("manager.fate.user.config", "{"
+      + "\"TABLE_CREATE,TABLE_DELETE,TABLE_RENAME,TABLE_ONLINE,TABLE_OFFLINE,NAMESPACE_CREATE,NAMESPACE_DELETE,NAMESPACE_RENAME,TABLE_TABLET_AVAILABILITY,SHUTDOWN_TSERVER\": 1,"
+      + "\"TABLE_BULK_IMPORT2\": 2,"
+      + "\"TABLE_COMPACT,TABLE_CANCEL_COMPACT,COMMIT_COMPACTION\": 4,"
```
COMMIT_COMPACTION will likely be the most frequently executed fate operation in a system.  A single TABLE_COMPACT operation can cause N COMMIT_COMPACTION operations to eventually run.  Also system compactions that occur automatically as data is ingested can cause COMMIT_COMPACTION ops to run.  Should probably move TABLE_COMPACT,TABLE_CANCEL_COMPACT elsewhere w/ the other table ops as they are much less frequent operations and are always user initiated.

I would recommend three default groups.   One for COMMIT_COMPACTION, one for SYSTEM_SPLIT, and one for everythnig else (I think everything else is the user initiated operations).  This will give dedicated resrouces to the the two system initated operations and keep them from intefering w/ the user initiated ops.

kevinrr888 commented:
Changes:
- Moved the resizing task from Fate.FatePoolsWatcher into FateExecutor. This cleaned up the code a bit getting rid of some getters in FateExecutor
- When looking at the new MANAGER_FATE_<META/USER>_CONFIG props again realized the description I wrote wasn't correct. Updated
- When tinkering with MANAGER_FATE_META_CONFIG and running the sunny day build realized that the config could be blank and still pass the sunny day build (meaning no FATE operations would be able to run for the Accumulo system tables). I assumed the sunny day build would catch something like this, but it doesn't appear that is the case. Perhaps we should create a sunny day test which tests fate operations on Accumulo system tables or maybe this would be caught in the full build and that is sufficient (I have not run the full build to see if that's true)... Regardless, this made me less confident in the correctness of my original attempt to write the META config (and looking at it again, I don't think it's correct). I changed the config to allow and expect all FateOperations, instead of trying to determine what subset would be allowed and expected for Accumulo system tables.
- Minor changes to PropertyTypeTest testTypeFATE_<META/USER>_CONFIG to make the test easier to understand
- Added new test to FatePoolsWatcherIT as described

@keith-turner - Could you take a look at these changes? Most of the changes to Fate and FateExecutor were copy-paste so those can be ignored. Mostly curious what you think about the changes to Property and the new test.

keith-turner commented on test/src/main/java/org/apache/accumulo/test/fate/FatePoolsWatcherIT.java:
```diff
@@ -357,6 +357,151 @@ protected void testIdleCountHistory(FateStore<PoolResizeTestEnv> store, ServerCo
+      // anything
+      Thread.sleep(fate.getPoolWatcherDelay().toMillis() + 1_000);
+
+      assertEquals(env.numWorkers.get(),
```
This is a good check that verifies things from the actualy running thread perspective, however it only verifies the sum.  In addition to this check maybe could do something like the following to verifiy what is in the store at a more granular level, if it would work.   The checks together would give a stronger verification.

```
      Map<Fate.FateOperation,Long> seenCounts =  store.list().filter(fateIdStatus -> fateIdStatus.getFateOperation().isPresent() && fateIdStatus.getFateReservation().isPresent()).collect(Collectors.groupingBy(fis->fis.getFateOperation().orElseThrow(), Collectors.counting()));
      Map<Fate.FateOperation,Long> expectedCounts = Map.of(fateOpFromPool1, (long)Math.min(numThreadsPool1, numSeedPool1),...);
      assertEquals(expectedCounts, seenCounts);
```

keith-turner commented:
> Perhaps we should create a sunny day test which tests fate operations on Accumulo system tables or maybe this would be caught in the full build and that is sufficient (I have not run the full build to see if that's true).

It does not have to be sunny day test, but we should probably have ITs that execute every table opertion against the system tables to ensure they behave as expected.  Would need to survey and see what coverage we currently have for this, I know there is some but not sure its complete.  Not something for this PR.

> @keith-turner - Could you take a look at these changes? Most of the changes to Fate and FateExecutor were copy-paste so those can be ignored. Mostly curious what you think about the changes to Property and the new test.

@kevinrr888 I looked over the changes and they look good. I made one comment about a possible improvement to the test that would be nice if its workable.

kevinrr888 commented:
@keith-turner
> I made one comment about a possible improvement to the test that would be nice if its workable.

That is a nice additional check. I added it throughout the tests where applicable. 

I also merged main into this branch. The only thing noteworthy of that is the new `SYSTEM_MERGE` operation that was recently added. In the default value of the new properties, I added this as shown for the config for `META` and `USER`:
```
         "{\"TABLE_CREATE,TABLE_DELETE,TABLE_RENAME,TABLE_ONLINE,TABLE_OFFLINE,NAMESPACE_CREATE,"
          + "NAMESPACE_DELETE,NAMESPACE_RENAME,TABLE_TABLET_AVAILABILITY,SHUTDOWN_TSERVER,"
          + "TABLE_BULK_IMPORT2,TABLE_COMPACT,TABLE_CANCEL_COMPACT,TABLE_MERGE,TABLE_DELETE_RANGE,"
          + "TABLE_SPLIT,TABLE_CLONE,TABLE_IMPORT,TABLE_EXPORT,SYSTEM_MERGE\": 4,"
          + "\"COMMIT_COMPACTION\": 4,\"SYSTEM_SPLIT\": 4}"
```
If we find that `SYSTEM_MERGE` (or any of the ops) need more resources, we can change this default.

> It does not have to be sunny day test, but we should probably have ITs that execute every table opertion against the system tables to ensure they behave as expected. Would need to survey and see what coverage we currently have for this, I know there is some but not sure its complete. Not something for this PR.

I'll open an issue for this

Merging this into main now
